File: PInvoke/Parser/ExpressionEvaluator.vb
' Copyright (c) Microsoft Corporation.  All rights reserved.

Imports System.Runtime.InteropServices

Namespace Parser

    ''' <summary>
    ''' Used to evaluate basic expressions encounter by the parser.  
    ''' </summary>
    ''' <remarks></remarks>
    Public Class ExpressionEvaluator
        Private _parser As New ExpressionParser()
        Private _opts As ScannerOptions

        Public Sub New()
            _opts = New ScannerOptions()
            _opts.HideComments = True
            _opts.HideNewLines = True
            _opts.HideWhitespace = True
            _opts.ThrowOnEndOfStream = False
        End Sub

        Public Function TryEvaluate(ByVal expr As String, <Out> ByRef result As ExpressionValue) As Boolean
            Dim list As List(Of Token) = Scanner.TokenizeText(expr, _opts)
            Return TryEvaluate(list, result)
        End Function

        Public Function TryEvaluate(ByVal list As List(Of Token), <Out> ByRef result As ExpressionValue) As Boolean
            Dim node As ExpressionNode = Nothing
            If Not _parser.TryParse(list, node) Then
                result = Nothing
                Return False
            End If

            Return TryEvaluate(node, result)
        End Function

        Public Function TryEvaluate(ByVal node As ExpressionNode, <Out> ByRef result As ExpressionValue) As Boolean
            If Not TryEvaluateCore(node) Then
                Return False
            End If

            result = DirectCast(node.Tag, ExpressionValue)
            Return True
        End Function

        Private Function TryEvaluateCore(ByVal node As ExpressionNode) As Boolean
            If node Is Nothing Then
                Return True
            End If

            ' Make sure that the left and right are evaluated appropriately
            If Not TryEvaluateCore(node.LeftNode) OrElse Not TryEvaluateCore(node.RightNode) Then
                Return False
            End If

            Select Case node.Kind
                Case ExpressionKind.BinaryOperation
                    Return TryEvaluateBinaryOperation(node)
                Case ExpressionKind.Leaf
                    Return TryEvaluateLeaf(node)
                Case ExpressionKind.NegativeOperation
                    Return TryEvaluateNegative(node)
                Case ExpressionKind.Cast
                    Return TryEvaluateCast(node)
                Case ExpressionKind.FunctionCall
                    Return TryEvaluateFunctionCall(node)
                Case ExpressionKind.NegationOperation
                    Return TryEvaluateNegation(node)
                Case ExpressionKind.List
                    Return TryEvaluateList(node)
            End Select
        End Function

        Protected Overridable Function TryEvaluateCast(ByVal node As ExpressionNode) As Boolean
            Return False
        End Function

        Protected Overridable Function TryEvaluateFunctionCall(ByVal node As ExpressionNode) As Boolean
            Return False
        End Function

        Protected Overridable Function TryEvaluateNegation(ByVal node As ExpressionNode) As Boolean
            Return False
        End Function

        Protected Overridable Function TryEvaluateList(ByVal node As ExpressionNode) As Boolean
            Return True
        End Function

        Protected Overridable Function TryEvaluateNegative(ByVal node As ExpressionNode) As Boolean
            node.Tag = -(DirectCast(node.LeftNode.Tag, ExpressionValue))
            Return True
        End Function

        Protected Overridable Function TryEvaluateLeaf(ByVal node As ExpressionNode) As Boolean
            Dim token As Token = node.Token
            If token.IsNumber Then
                Dim value As Object = Nothing
                If Not TokenHelper.TryConvertToNumber(node.Token, value) Then
                    Return False
                End If
                node.Tag = New ExpressionValue(value)
                Return True
            ElseIf token.TokenType = TokenType.TrueKeyword Then
                node.Tag = New ExpressionValue(True)
                Return True
            ElseIf token.TokenType = TokenType.FalseKeyword Then
                node.Tag = New ExpressionValue(False)
                Return True
            ElseIf token.IsCharacter Then
                Dim cValue As Char = "0"c
                If Not TokenHelper.TryConvertToChar(node.Token, cValue) Then
                    Return False
                End If
                node.Tag = New ExpressionValue(cValue)
                Return True
            ElseIf token.IsQuotedString Then
                Dim sValue As String = Nothing
                If Not TokenHelper.TryConvertToString(token, sValue) Then
                    Return False
                End If
                node.Tag = New ExpressionValue(sValue)
                Return True
            Else
                Return False
            End If
        End Function

        Protected Overridable Function TryEvaluateBinaryOperation(ByVal node As ExpressionNode) As Boolean
            Dim left As ExpressionValue = DirectCast(node.LeftNode.Tag, ExpressionValue)
            Dim right As ExpressionValue = DirectCast(node.RightNode.Tag, ExpressionValue)
            Dim result As ExpressionValue = Nothing
            Select Case node.Token.TokenType
                Case TokenType.OpDivide
                    result = left / right
                Case TokenType.OpGreaterThan
                    result = New ExpressionValue(left > right)
                Case TokenType.OpGreaterThanOrEqual
                    result = New ExpressionValue(left >= right)
                Case TokenType.OpLessThan
                    result = New ExpressionValue(left < right)
                Case TokenType.OpLessThanOrEqual
                    result = New ExpressionValue(left <= right)
                Case TokenType.OpMinus
                    result = left - right
                Case TokenType.OpModulus
                    result = left - ((left \ right) * right)
                Case TokenType.OpShiftLeft
                    result = left << CInt(right.Value)
                Case TokenType.OpShiftRight
                    result = left >> CInt(right.Value)
                Case TokenType.OpPlus
                    result = left + right
                Case TokenType.OpBoolAnd
                    result = left AndAlso right
                Case TokenType.OpBoolOr
                    result = left OrElse right
                Case TokenType.OpEquals
                    result = New ExpressionValue(left.Value.Equals(right.Value))
                Case TokenType.OpNotEquals
                    result = New ExpressionValue(Not (left.Value.Equals(right.Value)))
                Case TokenType.OpAssign
                    result = right
                Case TokenType.Ampersand
                    result = left And right
                Case TokenType.Pipe
                    result = left Or right
                Case Else
                    Debug.Fail("Unrecognized binary operation")
                    Return False
            End Select

            node.Tag = result
            Return node.Tag IsNot Nothing
        End Function

    End Class

End Namespace

File: PInvoke.Core/Parser/ExpressionEvaluator.cs
// Copyright (c) Microsoft Corporation.  All rights reserved.

using PInvoke.Parser.Enums;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Runtime.InteropServices;

namespace PInvoke.Parser
{
    /// <summary>
    /// Used to evaluate basic expressions encounter by the parser.
    /// </summary>
    public sealed class ExpressionEvaluator
    {
        private readonly ExpressionParser parser = new ExpressionParser();
        private readonly Dictionary<string, Macro> macroMap;
        private readonly ScannerOptions opts;

        public ExpressionEvaluator(Dictionary<string, Macro> macroMap = null)
        {
            this.macroMap = macroMap ?? new Dictionary<string, Macro>();
            opts = new ScannerOptions
            {
                HideComments = true,
                HideNewLines = true,
                HideWhitespace = true,
                ThrowOnEndOfStream = false
            };
        }

        public bool TryEvaluate(string expr, out ExpressionValue result)
        {
            var list = Scanner.TokenizeText(expr, opts);
            return TryEvaluate(list, out result);
        }

        public bool TryEvaluate(List<Token> list, out ExpressionValue result)
        {
            if (!parser.TryParse(list, out ExpressionNode node))
            {
                result = null;
                return false;
            }

            return TryEvaluate(node, out result);
        }

        public bool TryEvaluate(ExpressionNode node, out ExpressionValue result)
        {
            try
            {
                result = EvaluateCore(node);
                return true;
            }
            catch
            {
                result = null;
                return false;
            }
        }

        public ExpressionValue Evalaute(ExpressionNode node)
        {
            return EvaluateCore(node);
        }

        private ExpressionValue EvaluateCore(ExpressionNode node)
        {
            Contract.ThrowIfNull(node);

            switch (node.Kind)
            {
                case ExpressionKind.BinaryOperation:
                    return EvaluateBinaryOperation(node);
                case ExpressionKind.Leaf:
                    return EvaluateLeaf(node);
                case ExpressionKind.NegativeOperation:
                    return EvaluateNegative(node);
                case ExpressionKind.Cast:
                    return EvaluateCast(node);
                case ExpressionKind.FunctionCall:
                    return EvaluateFunctionCall(node);
                case ExpressionKind.NegationOperation:
                    return EvaluateNegation(node);
                case ExpressionKind.List:
                    return TryEvaluateList(node);
                default:
                    throw Contract.CreateInvalidEnumValueException(node.Kind);
            }
        }

        /// <summary>
        /// For a cast just return the value of the left node
        /// </summary>
        private ExpressionValue EvaluateCast(ExpressionNode node)
        {
            // TODO: why left here?  Shouldn't it be right? 
            return EvaluateCore(node.LeftNode);
        }

        private ExpressionValue EvaluateFunctionCall(ExpressionNode node)
        {
            bool value =
                node.Token.Value == "defined" &&
                node.LeftNode != null &&
                macroMap.ContainsKey(node.LeftNode.Token.Value);

            return ExpressionValue.Create(value);
        }

        private ExpressionValue EvaluateNegation(ExpressionNode node)
        {
            var value = EvaluateCore(node.LeftNode);
            return ExpressionValue.Create(!value.ConvertToBool());
        }

        private ExpressionValue TryEvaluateList(ExpressionNode node)
        {
            throw new Exception($"Cannot evaluate list");
        }

        private ExpressionValue EvaluateNegative(ExpressionNode node)
        {
            var exprValue = EvaluateCore(node.LeftNode);
            if (exprValue.IsFloatingPoint)
            {
                var value = exprValue.ConvertToDouble();
                return ExpressionValue.Create(-value);
            }
            else
            {
                var value = exprValue.ConvertToInteger();
                return ExpressionValue.Create(-value);
            }
        }

        private ExpressionValue EvaluateLeaf(ExpressionNode node)
        {
            var token = node.Token;
            if (token.IsNumber)
            {
                if (!TokenHelper.TryConvertToNumber(node.Token, out Number value))
                {
                    throw new Exception($"Can't convert token to number {node.Token}");
                }

                return ExpressionValue.Create(value);
            }
            else if (token.TokenType == TokenType.TrueKeyword)
            {
                return ExpressionValue.Create(true);
            }
            else if (token.TokenType == TokenType.FalseKeyword)
            {
                return ExpressionValue.Create(false);
            }
            else if (token.TokenType == TokenType.Word)
            {
                return EvaluateMacro(node);
            }
            else if (token.IsCharacter)
            {
                var cValue = '0';
                if (!TokenHelper.TryConvertToChar(node.Token, out cValue))
                {
                    throw new Exception($"Can't convert token to char {node.Token}");
                }
                return ExpressionValue.Create(cValue);
            }
            else if (token.IsQuotedString)
            {
                if (!TokenHelper.TryConvertToString(token, out string sValue))
                {
                    throw new Exception($"Can't convert token to string {node.Token}");
                }
                return ExpressionValue.Create(sValue);
            }
            else if (TokenHelper.IsKeyword(node.Token.TokenType))
            {
                return ExpressionValue.Create(1);
            }
            else
            {
                throw new Exception($"Unexpected leaf token {node.Token}");
            }
        }

        private ExpressionValue EvaluateMacro(ExpressionNode node)
        {
            Contract.Requires(node.Kind == ExpressionKind.Leaf);
            Contract.Requires(node.Token.TokenType == TokenType.Word);

            var value = default(ExpressionValue);
            if (macroMap.TryGetValue(node.Token.Value, out Macro m))
            {
                if (TokenHelper.TryConvertToNumber(m.Value, out Number numValue))
                {
                    value = ExpressionValue.Create(numValue);
                }
                else
                {
                    value = ExpressionValue.Create(1);
                }
            }
            else
            {
                value = ExpressionValue.Create(0);
            }

            return value;
        }

        private ExpressionValue EvaluateBinaryOperation(ExpressionNode node)
        {
            if (!TryConvertToBinaryOperator(node.Token.TokenType, out BinaryOperator op))
            {
                throw new Exception($"Invalid binary node {node.Token.TokenType}");
            }

            var left = EvaluateCore(node.LeftNode);
            var right = EvaluateCore(node.RightNode);
            return EvaluateBinaryOperation(op, left, right);
        }

        public static bool TryConvertToBinaryOperator(TokenType type, out BinaryOperator op)
        {
            switch (type)
            {
                case TokenType.OpDivide:
                    op = BinaryOperator.Divide;
                    break;
                case TokenType.OpGreaterThan:
                    op = BinaryOperator.GreaterThan;
                    break;
                case TokenType.OpGreaterThanOrEqual:
                    op = BinaryOperator.GreaterThanOrEqualTo;
                    break;
                case TokenType.OpLessThan:
                    op = BinaryOperator.LessThan;
                    break;
                case TokenType.OpLessThanOrEqual:
                    op = BinaryOperator.LessThanOrEqualTo;
                    break;
                case TokenType.OpMinus:
                    op = BinaryOperator.Subtract;
                    break;
                case TokenType.OpModulus:
                    op = BinaryOperator.Modulus;
                    break;
                case TokenType.OpShiftLeft:
                    op = BinaryOperator.ShiftLeft;
                    break;
                case TokenType.OpShiftRight:
                    op = BinaryOperator.ShiftRight;
                    break;
                case TokenType.OpPlus:
                    op = BinaryOperator.Add;
                    break;
                case TokenType.OpBoolAnd:
                    op = BinaryOperator.BooleanAnd;
                    break;
                case TokenType.OpBoolOr:
                    op = BinaryOperator.BooleanOr;
                    break;
                case TokenType.OpEquals:
                    op = BinaryOperator.BooleanEquals;
                    break;
                case TokenType.OpNotEquals:
                    op = BinaryOperator.BooleanNotEquals;
                    break;
                case TokenType.Ampersand:
                    op = BinaryOperator.BitwiseAnd;
                    break;
                case TokenType.Pipe:
                    op = BinaryOperator.BitwiseOr;
                    break;
                case TokenType.OpAssign:
                    op = BinaryOperator.Assign;
                    break;
                default:
                    op = BinaryOperator.Add;
                    return false;
            }

            return true;
        }

        /// <summary>
        /// Reference used for this implementation
        /// 
        /// https://gcc.gnu.org/onlinedocs/cpp/If.html#If
        /// 
        /// In summary calculations done using the widest type known to compiler which is long for 
        /// our implementation.
        /// </summary>
        public static ExpressionValue EvaluateBinaryOperation(BinaryOperator op, ExpressionValue left, ExpressionValue right)
        {
            long leftValue;
            bool leftBool;
            long rightValue;
            bool rightBool;

            try
            {
                leftValue = left.ConvertToLong();
                leftBool = left.ConvertToBool();
                rightValue = right.ConvertToLong();
                rightBool = right.ConvertToBool();
            }
            catch (Exception ex)
            {
                throw new Exception($"Unable to convert binary operands: {ex.Message}", ex);
            }

            ExpressionValue result;
            switch (op)
            {
                case BinaryOperator.Divide:
                    result = ExpressionValue.Create(leftValue / rightValue);
                    break;
                case BinaryOperator.GreaterThan:
                    result = ExpressionValue.Create(leftValue > rightValue);
                    break;
                case BinaryOperator.GreaterThanOrEqualTo:
                    result = ExpressionValue.Create(leftValue >= rightValue);
                    break;
                case BinaryOperator.LessThan:
                    result = ExpressionValue.Create(leftValue < rightValue);
                    break;
                case BinaryOperator.LessThanOrEqualTo:
                    result = ExpressionValue.Create(leftValue <= rightValue);
                    break;
                case BinaryOperator.Subtract:
                    result = ExpressionValue.Create(leftValue - rightValue);
                    break;
                case BinaryOperator.Modulus:
                    result = ExpressionValue.Create(leftValue % rightValue);
                    break;
                case BinaryOperator.ShiftLeft:
                    result = ExpressionValue.Create(leftValue << (int)rightValue);
                    break;
                case BinaryOperator.ShiftRight:
                    result = ExpressionValue.Create(leftValue >> (int)rightValue);
                    break;
                case BinaryOperator.Add:
                    result = ExpressionValue.Create(leftValue + rightValue);
                    break;
                case BinaryOperator.Multiply:
                    result = ExpressionValue.Create(leftValue * rightValue);
                    break;
                case BinaryOperator.BooleanAnd:
                    result = ExpressionValue.Create(leftBool && rightBool);
                    break;
                case BinaryOperator.BooleanOr:
                    result = ExpressionValue.Create(leftBool || rightBool);
                    break;
                case BinaryOperator.BooleanEquals:
                    result = ExpressionValue.Create(leftValue == rightValue);
                    break;
                case BinaryOperator.BooleanNotEquals:
                    result = ExpressionValue.Create(leftValue != rightValue);
                    break;
                case BinaryOperator.BitwiseAnd:
                    result = ExpressionValue.Create(leftValue & rightValue);
                    break;
                case BinaryOperator.BitwiseOr:
                    result = ExpressionValue.Create(leftValue | rightValue);
                    break;
                case BinaryOperator.Assign:
                    result = right;
                    break;
                default:
                    throw Contract.CreateInvalidEnumValueException(op);
            }

            return result;
        }
    }
}
